### PR TITLE
feat: implement basic error handling

### PR DESCRIPTION
**Pull request by jupyterjazz**
Implements basic error handling.
We can come up with more sophisticated error classes/types after.

### CODE REVIEW

File: finetuner/client/exception.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
+from typing import Optional
+
+
+class BaseError(Exception):
+    response = None
+    code = -1
+    message = "An unknown error occurred"
```
Comment (gmastrapas) [inline finetuner/client/exception.py:7]: single quotes

File: finetuner/client/exception.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
+from typing import Optional
+
+
+class BaseError(Exception):
```
Comment (gmastrapas) [inline finetuner/client/exception.py:4]: I would name this `FinetunerServerError` for now

File: finetuner/client/exception.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    response = None
+    code = -1
+    message = "An unknown error occurred"
+
+    def __init__(
+        self,
+        response: dict,
+        message: Optional[str] = None,
```
Comment (gmastrapas) [inline finetuner/client/exception.py:12]: setting default None values makes the code above more complex. Just set the class attribute values as default here:
```
class BaseError(Exception):

def __init__(
        self,
        response: dict,
        message: str = "An unknown error occurred",
        code: int = -1,
```

File: finetuner/client/base.py [hunk 4]
```diff
@@ -60,5 +67,8 @@ def _handle_request(
         """
         response = self._session.request(
             url=url, method=method, json=json_data, params=params, verify=False
-        ).json()
-        return response
+        )
+        if not response.ok:
+            self._handle_error_request(response)
```
Comment (gmastrapas) [inline finetuner/client/base.py:72]: Just do here:

```
raise BaseError(response=response.json(), message=response.reason, code=response.status_code)
```
no need to have an additional function

File: finetuner/run.py [hunk 5]
```diff
@@ -53,10 +53,7 @@ def save_model(self, path: str = FINETUNED_MODELS_DIR):
         :returns: A list of str object(s) that indicate the download path.
         """
         if self.status() != FINISHED:
-            # tell users that the run hasn't finished yet, thus they can't download
-            # the model.
-            # will implement a proper error-handling in a separate pr.
-            raise ValueError
+            raise Exception("The run needs to be finished in order to save the model.")
```
Comment (gmastrapas) [inline finetuner/run.py:56]: single quotes

**Review by gmastrapas — CHANGES_REQUESTED** [review]

**Comment by jupyterjazz** [comment]
Looks like this
servererror

**Comment by gmastrapas** [comment]
> Looks like this servererror

Hmm I would say let's do it like this:
```
finetuner.client.exception.FinetunerServerError: INTERNAL SERVER ERROR (500) - details
```

**Comment by jupyterjazz** [comment]
@gmastrapas 
> Hmm I would say let's do it like this:
> 
> ```
> finetuner.client.exception.FinetunerServerError: INTERNAL SERVER ERROR (500) - details
> ```

`Internal Server Error` is the `message` (i.e. `details`) here.

If we change the way you suggested, we'll get something like this:
> finetuner.client.exception.FinetunerServerError: INTERNAL SERVER ERROR (500) - Internal Server Error

**Comment by gmastrapas** [comment]
> @gmastrapas
> 
> > Hmm I would say let's do it like this:
> > ```
> > finetuner.client.exception.FinetunerServerError: INTERNAL SERVER ERROR (500) - details
> > ```
> 
> `Internal Server Error` is the `message` (i.e. `details`) here.
> 
> If we change the way you suggested, we'll get something like this:
> 
> > finetuner.client.exception.FinetunerServerError: INTERNAL SERVER ERROR (500) - Internal Server Error

Inside the body there must be a `detail` field

**Comment by jupyterjazz** [comment]
@gmastrapas good point!

This is the new one
newerror

**Review by gmastrapas — APPROVED** [review]
Looks good!

**Pull request merged**